src/ship/GunManager.h

```diff
@@ -0,0 +1,222 @@
+		// TODO: integrate this with a whole-ship cooling system
+		// Currently uses GunManager::m_coolingBoost
+		// float coolerOverclock = 1.0; // any boost to gun coolers from ship equipment
+		// float powerOverclock = 1.0; // boost to overall firerate (and corresponding increase in heat generation)
+
+		vector3f currentLead;
+		vector3d currentLeadPos;
+
+		WeaponData data;
+		SceneGraph::Model *model; // gun model, currently unused
+	};
+
+	struct WeaponMount {
+		StringName id;
+		SceneGraph::Tag *tag;     // Tag in the parent model that this weapon mount is attached to
+		vector2f gimbalLimitTan;  // tangent of gimbal limits
+		//SceneGraph::Model *model; // model to render for this weapon mount, if any
+		// TODO: enable/disable hardpoint based on ship configuration, i.e. landing/vtol/wings?
+	};
+
+	struct GroupState {
+		WeaponIndexSet weapons;				// Whic weapons are assigned to this group?
+		const Body *target;					// The target for this group, if any
+		bool firing;						// Is the group currently firing
+		bool fireWithoutTargeting;			// Can the group fire without a target/target not in gimbal range?
+		ConnectionTicket onTargetDestroyed;	// Unlock the target once it's been destroyed
+	};
+
+	GunManager() = default;
+
+	void Init(ModelBody *b);
+
+	void SaveToJson(Json &jsonObj, Space *space);
+	void LoadFromJson(const Json &jsonObj, Space *space);
+
+	void StaticUpdate(float deltaTime);
+
+	// ==========================================
+
+	// Add a weapon mount to this gun manager
+	void AddWeaponMount(StringName id, StringName tagName, vector2f gimbalLimitDegrees);
+	// Remove a weapon mount from this gun manager. This will fail if a weapon is still mounted.
+	void RemoveWeaponMount(StringName id);
+
+	// Attach a weapon to a specific mount
+	bool MountWeapon(StringName hardpoint, const WeaponData &data);
+	// Remove the attached weapon from a specific mount
+	void UnmountWeapon(StringName hardpoint);
+	// Check if a weapon is attached to a specific mount
+	bool IsWeaponMounted(StringName hardpoint) const;
+
+	const std::vector<WeaponState> &GetWeapons() const { return m_weapons; }
+
+	uint32_t GetNumWeapons() const { return m_weapons.size(); }
+	const WeaponState *GetWeaponState(uint32_t numWeapon) const { return m_weapons.size() > numWeapon ? &m_weapons[numWeapon] : nullptr; }
+	uint32_t GetWeaponIndexForHardpoint(StringName hardpoint) const;
+
+	// ==========================================
+
+	// For AI/NPC ship usage, combines all weapons into a single group
+	void SetupDefaultGroups();
+	void SetTrackingTarget(const Body *target) { return SetGroupTarget(0, target); }
+	void SetAllGroupsFiring(bool firing = false);
+
+	// For player weapon management
+
+	// Assign a specific weapon to a firing group
+	// Weapons are by default all assigned to group 0
+	void AssignWeaponToGroup(uint32_t numWeapon, uint32_t group);
+	// Remove a weapon group in its entirety
+	void RemoveGroup(uint32_t group);
```

Comment:
Are groups just integers? Seems a player might want a more descriptive name..
(Haven't looked at the later commits yet, apologies if this is handled in LUA)

Reply:
Per my comment above, groups are a low-level construct. Any names, etc. will be handled at a much higher level (as will sorting of weapons). The intended design is something like Elite Dangerous or Starsector; see this example from E:D (note that at current I plan to only support a single fire button):

![](https://oyster.ignimgs.com/mediawiki/apis.ign.com/elite-dangerous/2/23/Fire_groups.png)

Reply:
Ok, that was my secondary question whether the naming would be handled in a higher layer, thank you.

A single fire button? We currently have Primary and Secondary already..?

Reply:
We have a "trigger" input and a "weapon release" input (Gun / Missile) - while the system can certainly be expanded to support multiple gun triggers (or trigger stages), I'd much prefer to handle that at a higher level than the GunManager. It's easier to have four sets of groups for four fire inputs than it is to have four bitfields for every single group.

Reply:
I was going by the in-game configuration screen for weapons; didn't look into the code to see how those were implemented/handled. Sorry.

I'm confused and perhaps misunderstood or talking at cross-purposes..  but what does having the number of fire buttons have to do with the implementation in GunManager? I thought you mentioned the single fire-button restriction as a direct comment on the E:D interface..

Reply:
> I'm confused and perhaps misunderstood or talking at cross-purposes.. but what does having the number of fire buttons have to do with the implementation in GunManager? I thought you mentioned the single fire-button restriction as a direct comment on the E:D interface..

I did mention that as a direct comment on the screenshot posted - there's a bit of topic-wandering on my end.

The mention of the number of fire buttons related to the implementation was meant to elucidate my thoughts on how we could implement (potentially multiple) simultaneous weapon-fire buttons (i.e. Primary, Secondary, Tertiary, Missile Release; or Fire Group 1, Fire Group 2, etc.) without having to make the GunManager responsible for the number of weapon inputs the player will use.

Reply:
The WeaponManager has a simple interface : `Set(All)GroupFire()`.  Up to the UI to manage which button triggers this..?  

Anyway; this comment is getting extremely derailed here!!

The original question was answered long ago, thank you.